### PR TITLE
Inprocrefactor

### PR DESCRIPTION
**Pull request by harshjain2**
Introduced IDataCollectionTestCaseEventManager that will send events to In-Proc and Out-Of-Proc datacollectors.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/vstest/517)

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 1]
```diff
@@ -263,6 +320,22 @@ private void SetInProcDataCollectionDataInTestResult(TestResult testResult)
...
+        /// </summary>
+        /// <param name="testResult">
+        /// The test Result.
+        /// </param>
+        /// <param name="allowTestResultFlush">
+        /// The allow Test Result Flush.
+        /// </param>
+        private void SetAllowTestResultFlushInTestResult(TestResult testResult, bool allowTestResultFlush)
+        {
+            var testProperty = TestProperty.Register(id: nameof(allowTestResultFlush), label: nameof(allowTestResultFlush), category: string.Empty, description: string.Empty, valueType: typeof(bool), validateValueCallback: null, attributes: TestPropertyAttributes.None, owner: typeof(TestCase));
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Do we require a TestProperty.Register? Or reuse `FlushResultTestResultPoperty`?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs [hunk 2]
```diff
@@ -46,62 +60,31 @@ public InProcDataCollectionExtensionManager(string runSettings, ITestRunCache te
...
-            }
-            catch (Exception ex)
-            {
-                EqtTrace.Error("InProcDataCollectionExtensionManager: Error occured while Initializing the datacollectors : {0}", ex);
-            }
-            finally
-            {
-                this.IsInProcDataCollectionEnabled = this.inProcDataCollectors.Any();
-            }
+            dataCollectionTestCaseEventManager.TestCaseEnd += this.TriggerTestCaseEnd;
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs]: Are the events subscribed even when Initialize fails? What if there are no inproc data collectors enabled?

**Comment by codito** [comment]
Reviewed 21 of 21 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 249 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYQzmN81k_mZ7nPcf4:-KdYQzmOuAu2bv2jmTWL:b-4nso9y) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L249)):*
> ```Smalltalk
>         /// <summary>
>         /// Loads all the inproc data collector dlls
>         /// </summary>       
> ```

nit: remove extra space at end

---

*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 283 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYRKIQxWnuCfXmOc1W:-KdYRKIQxWnuCfXmOc1X:bqgrwp0) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L283)):*
> ```Smalltalk
>                 this.IsInProcDataCollectionEnabled = this.inProcDataCollectors.Any();
>             }
>         }        
> ```

nit: remove trailing space

---

*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/Interfaces/IDataCollectionTestCaseEventManager.cs, line 79 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYRkbR8_YevYCDObnk:-KdYRkbS_3UBcjs2MaVo:b-yk0npv) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/Interfaces/IDataCollectionTestCaseEventManager.cs#L79)):*
> ```Smalltalk
>         /// </param>
>         void RaiseTestResult(TestResultEventArgs e);
> ```

nit: remove extra lines

---

*[src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestCaseEventsHandler.cs, line 32 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYW--mhjku37MR_lXY:-KdYW--nhMo-AmAFR89_:beumv7) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestCaseEventsHandler.cs#L32)):*
> ```Smalltalk
>         /// The test Case Events.
>         /// </param>
>         public TestCaseEventsHandler(IDataCollectionTestCaseEventManager dataCollectionTestCaseEventManager, ITestCaseEventsHandler testCaseEvents)
> ```

It is not very clear why a ctor takes in an object of itself. How does the caller construct this object?

(I know this code was existing earlier to this PR, it will be good to understand this)

---

*[src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectionContext.cs, line 12 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYX_W7slpKJtx2qqoJ:-KdYX_W7slpKJtx2qqoK:b-seh9wg) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectionContext.cs#L12)):*
> ```Smalltalk
>     /// </summary>
> #if NET46
>     [Serializable] 
> ```

Why do we need Serializable? Please remove. If this data object is transferred over wire, please explicitly mark DataContract.

Same comment applies to other files.

---

*[src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/SessionEvents.cs, line 18 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYYoARjx1yVwxMZ4lH:-KdYYoAS5qbOXEROXXHg:b-8zq5p5) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/SessionEvents.cs#L18)):*
> ```Smalltalk
>     {
>         #region Constructor
>         public SessionStartEventArgs() : this(new DataCollectionContext(new SessionId(Guid.Empty)))
> ```

Please add a comment on when can user expect `SessionId` to be non-empty; what does it signify.

---

*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 132 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYZlSqTHU55YfqUk2t:-KdYZlSqTHU55YfqUk2u:b-55iq9v) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L132)):*
> ```Smalltalk
>         public void TriggerSessionStartShouldCallInProcDataCollector()
>         {
>             var inProcDataCollectionManager = new TestableInProcDataCollectionExtensionManager(this.settingsXml, null, this.mockDataCollectionTestCaseEventManager.Object);
> ```

We can save the `n` number of changes in future if we remove DRY and create this object in .ctor or TestInitialize :)

---

*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 167 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdY_24aCe4Q_x-LYlqQ:-KdY_24bT6_jjxoO__IF:b-42hoos) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L167)):*
> ```Smalltalk
>         {
>             var inProcDataCollectionManager = new TestableInProcDataCollectionExtensionManager(this.settingsXml, null, this.mockDataCollectionTestCaseEventManager.Object);
>             Assert.IsTrue(inProcDataCollectionManager.IsInProcDataCollectionEnabled, "InProcDataCollection must be enabled if runsettings contains inproc datacollectors.");
> ```

This condition is tested in all tests, why not just have one test for this.

---

*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 193 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdY_kdKb0mPOiAO9JEw:-KdY_kdKb0mPOiAO9JEx:b-g09hv5) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L193)):*
> ```Smalltalk
>             // random guid
>             testCase.Id = new Guid("3871B3B0-2853-406B-BB61-1FE1764116FD");
>             inProcDataCollectionManager.TriggerTestCaseStart(null, new TestCaseStartEventArgs(testCase));
> ```

Why are we testing these APIs? Isn't IDataCollectionEventsHandler getting called everywhere? These APIs should be made private.

---

*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/EventHandlers/TestCaseEventHandlerTests.cs, line 95 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYa5AJvu8kemdk8KCp:-KdYa5AJvu8kemdk8KCq:b-9755fa) ([raw file](https://github.com/microsoft/vstest/blob/26ae9ea6aab555d05a88b710e6725947c680c371/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/EventHandlers/TestCaseEventHandlerTests.cs#L95)):*
> ```Smalltalk
>         }
> 
>         internal class TestableInProcDataCollectionExtensionManager : InProcDataCollectionExtensionManager
> ```

Thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestCaseEventsHandler.cs, line 32 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYW--mhjku37MR_lXY:-KdZUBgsnn8XXBiYezfr:bx88byn) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestCaseEventsHandler.cs#L32)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

It is not very clear why a ctor takes in an object of itself. How does the caller construct this object?

(I know this code was existing earlier to this PR, it will be good to understand this)
</blockquote></details>

We can get an instance of ITestCaseEventsHandler from here
https://github.com/Microsoft/vstest/blob/c701070abb126853b58800cedd7554482471ac7d/src/Microsoft.TestPlatform.Common/Interfaces/Engine/TesthostProtocol/IExecutionManager.cs#L31

However, we are currently passing null for this.
https://github.com/Microsoft/vstest/blob/c701070abb126853b58800cedd7554482471ac7d/src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestHandler.cs#L125

TestCaseEventsHandler is a decorator that provides additional functionality of sending events to DataCollectors. 

As of today, test case level events are used only by data collectors, but that could be value for supporting host extensibility, where in a custom host provides its own implementation of ITestCaseEventsHandler. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 66 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdWHiVr4R5hbrMsSjIS-r1-66:-KdZVqbZuSe95cgFXMA_:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L66)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

Are the events subscribed even when Initialize fails? What if there are no inproc data collectors enabled?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 283 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYRKIQxWnuCfXmOc1W:-KdZVuRiPSNWFS6IK_Vl:bkqpeuq) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L283)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

nit: remove trailing space
</blockquote></details>

This doesn't exist.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 249 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYQzmN81k_mZ7nPcf4:-KdZVzdXiCQ0WYJY6reY:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L249)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

nit: remove extra space at end
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/Interfaces/IDataCollectionTestCaseEventManager.cs, line 79 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYRkbR8_YevYCDObnk:-KdZW0dlWgjPDo3krwmJ:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/Interfaces/IDataCollectionTestCaseEventManager.cs#L79)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

nit: remove extra lines
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 167 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdY_24aCe4Q_x-LYlqQ:-KdZYwzNKCIW_bsu-uPe:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L167)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

This condition is tested in all tests, why not just have one test for this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 132 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYZlSqTHU55YfqUk2t:-KdZZG7ksaPV52x2gff2:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L132)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

We can save the `n` number of changes in future if we remove DRY and create this object in .ctor or TestInitialize :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/SessionEvents.cs, line 18 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYYoARjx1yVwxMZ4lH:-KdZbg0YBmnHbS0YEj3R:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/SessionEvents.cs#L18)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

Please add a comment on when can user expect `SessionId` to be non-empty; what does it signify.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs, line 336 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdWHiVr4R5hbrMsSjIS-r1-336:-KdZy2izj3MLG8kbdpHo:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollectionExtensionManager.cs#L336)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

Do we require a TestProperty.Register? Or reuse `FlushResultTestResultPoperty`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectionContext.cs, line 12 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdYX_W7slpKJtx2qqoJ:-KdZy83tyaqaqVD0VchR:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectionContext.cs#L12)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

Why do we need Serializable? Please remove. If this data object is transferred over wire, please explicitly mark DataContract.

Same comment applies to other files.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Comment by harshjain2** [comment]
*[test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs, line 193 at r1](https://reviewable.io:443/reviews/microsoft/vstest/517#-KdY_kdKb0mPOiAO9JEw:-Kd_LBJt1Bc35f6TF_uE:b-896fix) ([raw file](https://github.com/microsoft/vstest/blob/be164fde4b3ce17e885d952c146d7654af241e60/test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/InProcDataCollectionExtensionManagerTests.cs#L193)):*
<details><summary><i>Previously, codito (Arun Mahapatra) wrote…</i></summary><blockquote>

Why are we testing these APIs? Isn't IDataCollectionEventsHandler getting called everywhere? These APIs should be made private.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/517)*

**Pull request merged**